url863d2ab553_themadcreator_delaunay_tgz-pJ8-org_delaunay_PoissonDiscSamplesDemoJ8/src/org/delaunay/algorithm/Triangulation.java

```diff
@@ -0,0 +1,507 @@
+	 * vertex.
+	 * 
+	 * We guarantee that a triangle will be located for each vertex by first
+	 * creating a "super triangle" that is at least twice as large as the bounds
+	 * of the vertices. If this triangulation will be used for point location,
+	 * you will want to call {@link #setKeepSuperTriangle(true)} so that points
+	 * outside the convex hull of vertices may also be located.
+	 * 
+	 * Basic incremental triangulation method inspired by Paul Bourke's notes
+	 * and psuedocode. See: {@link http://paulbourke.net/papers/triangulate/}
+	 * 
+	 * @throws InvalidVertexException
+	 *             if any two vertices have the same location or if any three
+	 *             points are co-linear.
+	 */
+    public void triangulate() throws InvalidVertexException {
+    	/*
+    	 * Reset triangulation state
+    	 */
+    	resetTriangulation();
+		
+		if (Iterables.isEmpty(inputVertices)) {
+			return;
+		}
+
+		/*
+		 * Determine the supertriangle.
+		 */
+		createSuperTriangle(inputVertices);
+
+		/*
+		 * Sort vertices using hilbert curve to linearize triangulation
+		 * performance.
+		 */
+		log.debug("Linearizing with Hilbert Space-filling Curve");
+		List<Vertex> sortedVertices = getHilbertSortedVertices(inputVertices);
+
+		/*
+		 * Add vertices one at a time, updating the triangulation as we go.
+		 */
+		log.debug("Building Triangulation");
+		for (Vertex vertex : sortedVertices) {
+			try {
+				addVertexToTriangulation(vertex);
+			} catch (InvalidVertexException e) {
+				if (vertexExceptionStrategy == VertexExceptionStrategy.THROW_EXCEPTION) {
+					throw e;
+				} else {
+					// ignore
+				}
+			}
+		}
+		
+		/*
+		 * Cleanup
+		 */
+		clearLocator();
+		if (!keepSuperTriangle) {
+			removeSuperTriangle();
+		}
+
+		log.debug("Triangulation Complete");
+	}
+
+	public void clear() {
+		resetTriangulation();
+		inputVertices = Sets.newLinkedHashSet();
+	}
+	
+	private void resetTriangulation() {
+		triangles = Sets.newLinkedHashSet();
+		vertices = Sets.newLinkedHashSet();
+		clearLocator();
+	}
+
+	private List<Vertex> getHilbertSortedVertices(Iterable<? extends Vertex> verts) {
+		Rectangle2D bbox = Vectors.boundingBox(Lists.newArrayList(superVerts));
+		ScaledHilbertIndex  hilbert = new ScaledHilbertIndex(hilbertOrder, bbox);
+		for (Vertex v : verts) {
+			v.setHilbertIndex(hilbert.toIndex(v.x, v.y));
+		}
+		List<Vertex> sortedVertices = Lists.newArrayList(verts);
+		Collections.sort(sortedVertices, new Comparator<Vertex>() {
+			public int compare(Vertex v1, Vertex v2) {
+				return v1.getHilbertIndex().compareTo(v2.getHilbertIndex());
+			}
+		});
+		return sortedVertices;
+	}
+
+	public void addVertexToTriangulation(Vertex vertex) throws InvalidVertexException {
+		Collection<Triangle> toRemove = null, toAdd = null;
+
+		try {
+			/*
+			 * Get the set of triangles for which the vertex lies in its
+			 * circumcircle.
+			 */
+			toRemove = getCircumcircleTriangles(vertex);
+		} catch (NonDelaunayException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)